### PR TITLE
Push payload message to kafka topic #SB-11993

### PR DESCRIPTION
**Pull request by deveshMantra**

### CODE REVIEW

File: src/app/helpers/environmentVariablesHelper.js [hunk 1]
```diff
@@ -71,10 +71,12 @@ let envVariables = {
   sunbird_learner_service_health_status: 'true',
   sunbird_content_service_health_status: 'true',
   sunbird_portal_cassandra_db_health_status: 'true',
-  sunbird_portal_player_cdn_url: env.sunbird_portal_player_cdn_url
+  sunbird_portal_player_cdn_url: env.sunbird_portal_player_cdn_url,
+  KAFKA_HOST: process.env.sunbird_kafka_host || 'localhost:9092',
```
Comment (anoophm) [inline src/app/helpers/environmentVariablesHelper.js]: sunbird_kafka_host: process.env.sunbird_kafka_host,

File: src/app/helpers/environmentVariablesHelper.js [hunk 2]
```diff
@@ -71,10 +71,12 @@ let envVariables = {
...
+  KAFKA_TOPIC: process.env.sunbird_topic || "test"
```
Comment (anoophm) [inline src/app/helpers/environmentVariablesHelper.js]: sunbird_state_sso_topic: process.env.sunbird_state_sso_topic

File: src/app/helpers/kafkaHelperService.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+const kafka = require('kafka-node');
+const _ = require('lodash');
+const envHelper = require('./environmentVariablesHelper');
+
+const client = new kafka.KafkaClient({
+  kafkaHost: envHelper.KAFKA_HOST,
```
Comment (anoophm) [inline src/app/helpers/kafkaHelperService.js]: change variable name

File: src/app/helpers/kafkaHelperService.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+});
+
+producer.on('error', function (error) {
+  console.error("Errored at kafka", error)
+});
+
+const KafkaService = {
+
+  // send data to kafka
+  sendRecord: (data, kafkaTopic, callback = () => {
```
Comment (anoophm) [inline src/app/helpers/kafkaHelperService.js]: implement with promise, we need to move away from callback

File: src/app/helpers/kafkaHelperService.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      {
+        topic: kafkaTopic,
+        messages: data
+      }
+    ];
+    // Send record to Kafka
+    producer.send(record, callback)
+  }
+};
+
```
Comment (anoophm) [inline src/app/helpers/kafkaHelperService.js]: create a initialize method and export it. Call init method in server.js

File: src/app/helpers/kafkaHelperService.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+const kafka = require('kafka-node');
+const _ = require('lodash');
+const envHelper = require('./environmentVariablesHelper');
+
+const client = new kafka.KafkaClient({
```
Comment (anoophm) [inline src/app/helpers/kafkaHelperService.js]: move this inside initialize function

File: src/app/helpers/ssoHelper.js [hunk 7]
```diff
@@ -138,4 +140,50 @@ const handleGetUserByIdError = (error) => {
   }
   throw error.error || error.message || error;
 }
-module.exports = { keycloak, verifySignature, verifyToken, fetchUserWithExternalId, createUser, createSession, updatePhone, updateRoles };
+
+  // creating payload for kafka service
+const getDetails = (sessionDetails) => {
```
Comment (anoophm) [inline src/app/helpers/ssoHelper.js]: getKafkaPayloadData

File: src/app/helpers/ssoHelper.js [hunk 8]
```diff
@@ -138,4 +140,50 @@ const handleGetUserByIdError = (error) => {
...
+      'orgExternalId': jwtPayload.school_id || '',
+      'roles': jwtPayload.roles || [],
+      'userId': userDetails.id || '',
+      'organisations': userDetails.organisations || [],
+      'firstName': userDetails.firstName
+    }
+  }
+};
+
+const ssoLogin = async (req) => {
```
Comment (anoophm) [inline src/app/helpers/ssoHelper.js]: sendStateSsoKafkaMessage

File: src/app/package.json [hunk 9]
```diff
@@ -115,6 +116,7 @@
     "sb_cache_manager": "file:libs/sb_cache_manager",
     "sb_telemetry_util": "file:libs/sb_telemetry_util",
     "ua-parser-js": "^0.7.13",
+    "uniqid": "^5.0.3",
```
Comment (anoophm) [inline src/app/package.json]: why this is needed, we can use "uuid": "^3.1.0",

File: src/app/routes/ssoRoutes.js [hunk 10]
```diff
@@ -109,8 +109,9 @@ module.exports = (app) => {
   })
 
   app.get(successUrl, async (req, res) => { // to support mobile sso flow
-    res.status(200).sendFile('./success_loader.html', {root: __dirname })
-  })
+    await ssoLogin(req);
```
Comment (anoophm) [inline src/app/routes/ssoRoutes.js]: no need to await, this shouldnt affect sso login. If it fails sso also fails

File: src/app/helpers/kafkaHelperService.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  producer.on('ready', function () {
+    console.log('Kafka Producer is connected and ready.');
+  });
+
+  producer.on('error', function (error) {
+    console.error("Errored at kafka", error)
+  });
+}
+
+if (envHelper.sunbird_kafka_host && envHelper.sunbird_state_sso_topic) {
```
Comment (anoophm) [inline src/app/helpers/kafkaHelperService.js]: if (envHelper.sunbird_kafka_host)

File: src/app/helpers/kafkaHelperService.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  init();
+}
+
+const KafkaService = {
+
+  // send data to kafka
+  sendMessage: (data, kafkaTopic, callback = () => {
+  }) => {
+
+    if (!envHelper.sunbird_kafka_host && !envHelper.sunbird_state_sso_topic) {
```
Comment (anoophm) [inline src/app/helpers/kafkaHelperService.js]: if (!envHelper.sunbird_kafka_host)

File: src/app/helpers/ssoHelper.js [hunk 13]
```diff
@@ -138,4 +139,48 @@ const handleGetUserByIdError = (error) => {
...
+      'orgExternalId': jwtPayload.school_id || '',
+      'roles': jwtPayload.roles || [],
+      'userId': userDetails.id || '',
+      'organisations': userDetails.organisations || [],
+      'firstName': userDetails.firstName
+    }
+  }
+};
+
+const sendStateSsoKafkaMessage = async (req) => {
```
Comment (rajeevsathish) [inline src/app/helpers/ssoHelper.js]: it should be sendStateSSOKafkaMessage

**Pull request merged**